site-src/geps/gep-709.md

```diff
@@ -0,0 +1,260 @@
+ensure that this feature is safe, we need to enforce a handshake mechanism that
+requires resources in both namespaces to agree to this reference. To accomplish
+that, a new ReferencePolicy resource should be introduced.
+
+![Reference Policy](images/709-referencepolicy.png)
+
+With this model, Routes would be able to directly reference Routes and Services
+in other namespaces. These references would only be considered valid if a
+ReferencePolicy in the target namespace explicitly allowed it. 
+
+The following example shows how a HTTPRoute in namespace foo could reference
+a Service in namespace bar. In this example a ReferencePolicy in the bar
+namespace explicitly allows references to Services from HTTPRoutes in the foo
+namespace.
+
+```yaml
+kind: HTTPRoute
+metadata:
+  name: foo
+  namespace: foo
+spec:
+  rules:
+  - matches:
+    - path: /bar
+    forwardTo:
+      backend:
+      - name: bar
+        namespace: bar
+---
+kind: ReferencePolicy
+metadata:
+  name: bar
+  namespace: bar
+spec:
+  from:
+  - group: networking.gateway.k8s.io
+    kind: HTTPRoute
+    namespace: foo
+  to:
+  - group: core
+    kind: Service
+```
+
+### API
+This proposed API is fairly straightforward, but comes with a couple notable
+decisions:
+
+1. Each ReferencePolicy only supports a single From and To section. Additional
+   trust relationships can be modeled with additional ReferencePolicy resources.
+1. Resource names are intentionally excluded from this policy for simplicity and
+   because they rarely provide any meaningful protection.
+1. A single Namespace is allowed per "From" struct. Although a selector would be
+   more powerful it may encourage unnecessarily insecure configuration.
+
+```go
+// ReferencePolicy identifies cross namespace relationships that are trusted for
+// Gateway API.
+type ReferencePolicy struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the desired state of ReferencePolicy.
+    Spec ReferencePolicySpec `json:"spec,omitempty"`
+}
+
+
+// ReferencePolicySpec identifies a cross namespace relationship that is trusted
+// for Gateway API.
+type ReferencePolicySpec struct {
+    // From describes the trusted namespaces and kinds that can reference the
+    // resources described in "To".
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinItems=1
+    From []ReferencePolicyFrom `json:"from"`
+
+    // To describes the resources that may be referenced by the resources
+    // described in "From".
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinItems=1
+    To []ReferencePolicyTo `json:"to"`
+}
+
+// ReferencePolicyFrom describes trusted namespaces and kinds.
+type ReferencePolicyFrom struct {
+    // Group is the group of the referrent.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    Group string `json:"group"`
```

Comment:
Because all three of these values are required, each `from` reference must fully specify a group, kind, and namespace. This is nice for implementers because it's very straightforward, but will this be flexible enough, or should we consider label selectors here? 

The downside of including label selectors is that you then need two sets of fields that are both optional, so the zero values of the fields and slices of this struct really matter. This way is much simpler.

Reply:
I've been tempted to introduce selectors here but so far have avoided it due to the increased complexity they add. In v1alpha1 I think we have had a few too many selectors and it became difficult to keep track of everything. Direct references can still be confusing, but I think they're generally easier to understand. One of my primary concerns with selectors is that their default is so open. Interpreted literally, an empty selector would accept references from all namespaces. 

So all of that to say that I agree with you. If we really _need_ selectors, we can always consider adding them in the future.